### PR TITLE
Enable sql function ifnull, nullif and isnull

### PR DESCRIPTION
**Pull request by harold-wang**
Issue #950 , if available:

Description of changes:
Add flow control functions ifNull, nullIf and isNull (#950)

* Init
* UT
* IT

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 1]
```diff
@@ -82,4 +102,66 @@ private static FunctionResolver isNotNull() {
...
+   * @return v2 if v1 is null
+   */
+  public static ExprValue exprIfNull(ExprValue v1, ExprValue v2) {
+    if (v1.isNull()) {
+      return v2;
+    } else if (v1.isMissing()) {
+      return v2;
+    } else {
+      return v1;
+    }
```
Comment (chloe-zh) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:143]: Not a problem: combine the isNull and isMissing conditions?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 2]
```diff
@@ -82,4 +102,66 @@ private static FunctionResolver isNotNull() {
...
+      return v1;
+    } else if (v2.isNull()) {
+      return v1;
+    } else if (v2.isMissing()) {
+      return v1;
+    } else if (v1.value().equals(v2.value())) {
+      return LITERAL_NULL;
+    } else {
+      return v1;
+    }
```
Comment (chloe-zh) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:164]: combine conditions lines to more readable like above?

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java [hunk 3]
```diff
@@ -79,4 +104,93 @@ public void is_not_null_predicate() {
...
+  }
+
+  @Test
+  public void test_null_if_predicate() {
+    Expression v1 = dsl.literal(100);
+    Expression v2 = dsl.literal(200);
+    FunctionExpression result = dsl.nullif(v1, v2);
+    assertEquals(v1.valueOf(valueEnv()), result.valueOf(valueEnv()));
+
+    System.out.println("debug : " + result.valueOf(valueEnv()));
```
Comment (chloe-zh) [inline core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java:138]: remove this line?

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java [hunk 4]
```diff
@@ -79,4 +104,93 @@ public void is_not_null_predicate() {
...
+
+    v1 = dsl.literal(100);
+    v2 = dsl.literal(100);
+
+    result = dsl.nullif(v1, v2);
```
Comment (chloe-zh) [inline core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java:143]: remove the above line; compare the result with expected result?

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import org.elasticsearch.search.SearchHits;
+import org.json.JSONObject;
+
+import org.junit.Assume;
+import org.junit.Test;
+
+
+/**
+ * Created by allwefantasy on 8/25/16.
+ */
```
Comment (chloe-zh) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:45]: remove these lines?

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+public class FlowConrolFunctionIT extends SQLIntegTestCase {
```
Comment (chloe-zh) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:46]: Typo: Control

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+  @Override
+  public void init() throws Exception {
+    super.init();
+    TestUtils.enableNewQueryEngine(client());
+    loadIndex(Index.BANK);
+    loadIndex(Index.ACCOUNT);
+    loadIndex(Index.ONLINE);
+    loadIndex(Index.DATE);
+  }
```
Comment (chloe-zh) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:56]: It's fine to load only the account index and remove others since only the account index is used in this IT class. Could you also add some test cases over the null/missing values. The index `bank_with_null_values.json` might be helpful since it contains a couple of null or missing values in some fields.

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    final XContentParser parser = XContentFactory.xContent(XContentType.JSON).createParser(
+        NamedXContentRegistry.EMPTY,
+        LoggingDeprecationHandler.INSTANCE,
+        rsp);
+    return SearchResponse.fromXContent(parser).getHits();
+  }
+
+  private JSONObject executeJdbcRequest(String query) {
+    return new JSONObject(executeQuery(query, "jdbc"));
+  }
```
Comment (chloe-zh) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:120]: Could you use the `executeQuery` method inherited from the parent class `SQLIntegTestCase` to execute SQL queries, it will enforce to JDBC format by default. For the assertion, you can check out the `verifySchema`, `verifyDataRows` etc. methods in `MatcherUtils`. For example:
```
JSONObject result = executeQuery(SELECT ISNULL(lastname) AS name ...);
verifySchema(result, schema("ISNULL(lastname)"), "name" , "boolean");
verifyDataRows(...);
```
You can also checkout the IT cases in `DataTimeFunctionIT.java` for reference.

**Comment by chloe-zh** [comment]
Could you also update the user manual here: https://github.com/opendistro-for-elasticsearch/sql/blob/develop/docs/user/dql/functions.rst#conditional-functions

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java (hunk 7 above)
Comment (harold-wang) [reply integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:56]: Got it, thanks

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java (hunk 6 above)
Comment (harold-wang) [reply integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java:46]: Thanks, to fix in next commit shortly.

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java (hunk 3 above)
Comment (harold-wang) [reply core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java:127]: Thanks, to remove it in next commit.

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java (hunk 4 above)
Comment (harold-wang) [reply core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java:132]: got it, thx

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java (hunk 5 above)
Comment (harold-wang) [reply integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java]: Got it, thx.

File: legacy/src/main/antlr/OpenDistroSqlLexer.g4 [hunk 9]
```diff
@@ -171,6 +171,7 @@ MONTH:                              'MONTH';
 MONTHNAME:                          'MONTHNAME';
 MULTIPLY:                           'MULTIPLY';
 NOW:                                'NOW';
+NULLIF:                             'NULLIF';
```
Comment (dai-chen) [inline legacy/src/main/antlr/OpenDistroSqlLexer.g4:174]: No need to touch this file since this is legacy grammar file?

File: integ-test/src/test/resources/correctness/queries/flowcontrol.txt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+SELECT IFNULL(AvgTicketPrice, AvgTicketPrice * 2.0) FROM kibana_sample_data_flights
```
Comment (dai-chen) [inline integ-test/src/test/resources/correctness/queries/flowcontrol.txt:1]: To simplify, you can move these tests to https://github.com/opendistro-for-elasticsearch/sql/blob/develop/integ-test/src/test/resources/correctness/expressions/conditionals.txt. In this way, you don't need a `FROM` clause and test index. The only literal not supported is `NULL` but you've figured out the workaround (1/0).

**Review by dai-chen — COMMENTED** [review]
Please add doctest here: https://github.com/opendistro-for-elasticsearch/sql/blob/develop/docs/user/dql/functions.rst#conditional-functions

**Comment by dai-chen** [comment]
Btw, I missed `IF` function, could you add it too? You can do it same here or a separate PR. Thanks!

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java (hunk 8 above)
Comment (harold-wang) [reply integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowConrolFunctionIT.java]: Got it, thanks

File: integ-test/src/test/resources/correctness/queries/flowcontrol.txt (hunk 10 above)
Comment (harold-wang) [reply integ-test/src/test/resources/correctness/queries/flowcontrol.txt]: done.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 11]
```diff
@@ -82,4 +106,62 @@ private static FunctionResolver isNotNull() {
             .collect(
                 Collectors.toList()));
   }
+
+  private static FunctionResolver ifNull() {
+    FunctionName functionName = BuiltinFunctionName.IFNULL.getName();
+    List<ExprType> typeList = ExprCoreType.coreTypes();
+    typeList.add(UNKNOWN);
```
Comment (harold-wang) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:113]: Chen, we have to add UNKNOWN, otherwise, we will see the "UNKNOWN" type missing error, I think it is NONE type.
est_nullif_predicate()
com.amazon.opendistroforelasticsearch.sql.exception.ExpressionEvaluationException: nullif function expected {[BYTE,BYTE],[SHORT,SHORT],[INTEGER,INTEGER],[LONG,LONG],[FLOAT,FLOAT],[DOUBLE,DOUBLE],[BOOLEAN,BOOLEAN],[STRING,STRING],[TIMESTAMP,TIMESTAMP],[DATE,DATE],[TIME,TIME],[DATETIME,DATETIME],[INTERVAL,INTERVAL],[STRUCT,STRUCT],[ARRAY,ARRAY]}, but get [UNKNOWN,UNKNOWN]
	at com.amazon.opendistroforelasticsearch.sql.expression.function.FunctionResolver.resolve(FunctionResolver.java:46)
	at com.amazon.opendistroforelasticsearch.sql.expression.function.BuiltinFunctionRepository.resolve(BuiltinFunctionRepository.java:44)
	at com.amazon.opendistroforelasticsearch.sql.expression.function.BuiltinFunctionRepository.compile(BuiltinFunctionRepository.java:30)
	at com.amazon.opendistroforelasticsearch.sql.expression.DSL.function(DSL.java:500)
	at com.amazon.opendistroforelasticsearch.sql.expression.DSL.nullif(DSL.java:526)
	at com.amazon.opendistroforelasticsearch.sql.expression.operator.predicate.UnaryPredicateOperatorTest.test_nullif_predicate(UnaryPredicateOperatorTest.java:138)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)


test_isnull_predicate()
com.amazon.opendistroforelasticsearch.sql.exception.ExpressionEvaluationException: isnull function expected {[BYTE],[SHORT],[INTEGER],[LONG],[FLOAT],[DOUBLE],[BOOLEAN],[STRING],[TIMESTAMP],[DATE],[TIME],[DATETIME],[INTERVAL],[STRUCT],[ARRAY]}, but get [UNKNOWN]

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/correctness/CorrectnessIT.java [hunk 12]
```diff
@@ -52,7 +52,8 @@
  * Correctness integration test by performing comparison test with other databases.
  */
 @ESIntegTestCase.SuiteScopeTestCase
-@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE, numDataNodes = 3, supportsDedicatedMasters = false, transportClientRatio = 1)
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE, numDataNodes = 3,
```
Comment (penghuo) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/correctness/CorrectnessIT.java:55]: what is the difference?

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    loadIndex(Index.ACCOUNT);
+    loadIndex(Index.BANK_WITH_NULL_VALUES);
+  }
+
+  @Test
+  public void ifnullShouldPassJDBC() throws IOException {
+    JSONObject response = executeJdbcRequest(
+        "SELECT IFNULL(lastname, 'unknown') AS name FROM " + TEST_INDEX_ACCOUNT
+            + " GROUP BY name");
+    assertEquals("IFNULL(lastname, \'unknown\')", response.query("/schema/0/name"));
```
Comment (penghuo) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java:58]: could you use verifySchema and verifyDataRows to get rid of jsonpath access?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/BuiltinFunctionName.java [hunk 14]
```diff
@@ -138,6 +138,9 @@
    */
   IS_NULL(FunctionName.of("is null")),
   IS_NOT_NULL(FunctionName.of("is not null")),
+  IFNULL(FunctionName.of("ifnull")),
+  NULLIF(FunctionName.of("nullif")),
+  ISNULL(FunctionName.of("isnull")),
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/BuiltinFunctionName.java:143]: why we need ISNULL?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 11 above)
Comment (penghuo) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:113]: The null value doesn't have a type. the problem only related to NULL literal test, e.g. ifnull(NULL), do we need to support NULL literal?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 15]
```diff
@@ -82,4 +106,62 @@ private static FunctionResolver isNotNull() {
...
+
+    List<SerializableFunction<FunctionName, org.apache.commons.lang3.tuple.Pair<FunctionSignature,
+            FunctionBuilder>>> functionsOne = typeList.stream().map(v ->
+            impl((UnaryPredicateOperator::exprIfNull), v, v, v))
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:117]: do we require the paramater of ifnull have same type? ifnull(expr1, expr2), the type of expr1 and expr2 must be same, right?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 16]
```diff
@@ -82,4 +106,62 @@ private static FunctionResolver isNotNull() {
...
+
+  /** null if v1 equls to v2.
+   * @param v1 varable 1
+   * @param v2 varable 2
+   * @return null if v1 equls to v2
+   */
+  public static ExprValue exprNullIf(ExprValue v1, ExprValue v2) {
+    if (v1.isNull() || v1.isMissing() || v2.isNull() || v2.isMissing()) {
+      return v1;
+    } else if (v1.value().equals(v2.value())) {
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:160]: could we use the v1.equals(v2) directly?

**Comment by penghuo** [comment]
* Please add the documentation accordingly
* Remove the leading "*" in the PR title

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 16 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:160]: yes, thx

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 15 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:117]: yes, expr1 and expr2 must be same, otherwise return value is correct, but return type could be wrong.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/BuiltinFunctionName.java (hunk 14 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/BuiltinFunctionName.java:143]: isnull(expr1, expr2), is_null is for is null(with space between is and null).

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 16 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:160]: public class Main {
  public static void main(String[] args) {
    String myStr1 = null;
    System.out.println("debug : " + myStr1.equals("test"));
  }
}
Exception in thread "main" java.lang.NullPointerException	at Main.main(Main.java:4)
if v1 is null, using the v1.equals(v2) directly will cause exception.

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java (hunk 13 above)
Comment (harold-wang) [reply integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java:61]: Done, thanks

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 11 above)
Comment (dai-chen) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:119]: @harold-wang I think you can remove null literal test from your UT and code here. This null literal support is missing for all expressions. We need to handle it in one place.

Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:119]: Let me remove null literal, thanks @penghuo @dai-chen

Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:119]: @dai-chen remove the UNKNOWN(null) type will make isnull() function totally useless, how about to keep it for now and remove it once you can enable it in generic way, thx.

Comment (dai-chen) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:119]: @harold-wang I don't think so. If we don't consider NULL literal, you won't get any UNKNOWN type input. For example, in your IT, the value of 1/0 is null but its type is integer. Could you post an example in which test case this is required?

Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:119]: All IT passed without UNKNOWN type, but UT has some coverage issue: what make sense to me:

  public static ExprValue exprIsNull(ExprValue v) {
    if (!v.isNull() && !v.isMissing()) {
      return LITERAL_FALSE;  // added by Harold, how can we cover here without UNKNOWN type.
    } else {
      return LITERAL_TRUE;
    }
  }

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 17]
```diff
@@ -74,6 +90,20 @@ private static FunctionResolver isNull() {
                 Collectors.toList()));
   }
 
+  private static FunctionResolver isNull() {
```
Comment (chloe-zh) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:93]: From my point of view, this method should be identical to the `is_Null()` method above. The setting of `UNKNOWN` type is not necessary

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java [hunk 18]
```diff
@@ -64,10 +82,9 @@ public ExprValue not(ExprValue v) {
     }
   }
 
-  private static FunctionResolver isNull() {
-
+  private static FunctionResolver is_Null(BuiltinFunctionName funcName) {
```
Comment (dai-chen) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java:85]: Could you keep original name `isNull`?

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java [hunk 19]
```diff
@@ -59,18 +65,37 @@ public void test_not_missing() {
...
     assertEquals(LITERAL_TRUE, expression.valueOf(valueEnv()));
   }
 
   @Test
-  public void is_not_null_predicate() {
+  public void test_isnull_predicate() {
+    ArrayList<Expression> exprValueArrayList = new ArrayList<>();
+    exprValueArrayList.add(dsl.literal("test"));
+    exprValueArrayList.add(dsl.literal(100));
+    exprValueArrayList.add(dsl.literal(""));
```
Comment (dai-chen) [inline core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java:83]: np: you can refactor this in your PR by JUnit's parameterized test. Example: https://github.com/opendistro-for-elasticsearch/sql/blob/develop/core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/function/WideningTypeRuleTest.java#L103

File: integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.search.SearchHits;
+import org.json.JSONObject;
+
+import org.junit.Assume;
+import org.junit.Test;
+
+public class FlowControlFunctionIT extends SQLIntegTestCase {
```
Comment (dai-chen) [inline integ-test/src/test/java/com/amazon/opendistroforelasticsearch/sql/sql/FlowControlFunctionIT.java:43]: np: noticed we use "conditional" in doc and comparison test. could you rename to `ConditionalIT` instead?

File: docs/experiment/ppl/functions/condition.rst [hunk 21]
```diff
@@ -67,3 +67,78 @@ Example, the account 13 doesn't have email field::
...
+------
+
+Description
+>>>>>>>>>>>
+
+Usage: ifnull(field1, field2) return field2 if field1 is null.
+
+Argument type: all the supported data type, (NOTE : if two parametershas different type, return result is correct, but return type could be wrong)
+
+Return type: any
```
Comment (dai-chen) [inline docs/experiment/ppl/functions/condition.rst:80]: Fo argument type, it should be "if two parameters have different type, you will fail semantic check"?

For return type, to avoid confusion, could you elaborate a little more than "any"?

Same for other function docs in PPL and SQL added.

**Review by dai-chen — APPROVED** [review]
Just several minor comments. Thanks for your changes!

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java (hunk 11 above)
Comment (harold-wang) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperator.java]: UNKNOWN is removed now.

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java (hunk 19 above)
Comment (harold-wang) [reply core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/operator/predicate/UnaryPredicateOperatorTest.java]: Done,thx

**Review by chloe-zh — APPROVED** [review]
